sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala

```diff
@@ -981,17 +988,26 @@ object SparkSession extends Logging {
    * @since 2.0.0
    */
   def setActiveSession(session: SparkSession): Unit = {
-    activeThreadSession.set(session)
+    if (session != getActiveSession.get && getActiveSession.isDefined) {
+      numActiveSessions.getAndIncrement
```

Comment:
The problem I see here is: if I have 2 sessions, and I set one as active then set another. If I keep doing this then the count here will be wrong.

I don't have a good idea to track the last alive session. Even if we can, users may want to create more sessions later and not stop the SparkContext.

The `SparkSession` should be light-weighted, what's the memory leak you observed before?

Reply:
Another problem is that the current behaviour is clearly documented. It doesn't look particularly wrong either:

> Stop the underlying `SparkContext`

We're trying to make a behaviour change just based on the new design choice.
Shell we at least keep the compatibility via a switch?

Reply:
yea, the right way to "free" a session is to leave it and wait for it to be GCed. If there is something that can't be GCed, it's a memory leak and we should fix it.

Reply:
@cloud-fan @HyukjinKwon - thanks for the thoughts. 

@cloud-fan - The memory leak is detailed here https://github.com/apache/spark/pull/24807#issuecomment-500871806.

@HyukjinKwon - I actually think that despite the fact that the current behavior is clearly documented, it actually doesn't make sense and is error prone. I detailed an example here https://github.com/apache/spark/pull/24807#issuecomment-500871806 of how this could bit us. 

Is there a world where a user should be able to call `.stop()` on a session and forcibly invalidate every other potential session (by killing the context)?

Reply:
If the leak is problem, we should fix it rather than changing the behaviour. It is documented and users are relying on this behaviour
We can understand `.stop()` is like `.stopContext()`, no?

I don't think we should just change without guarding. All other projects related to Spark such as Zeppelin would need to revisit their behaviour about how to stop, and it would make it difficult them to support multiple Spark versions for instance.